### ISSUE TITLE
HTTP response streaming and buffer issue?

### ISSUE THREAD

**Issue by thojanssens** [ericmj/mint#219]
I am trying to read a response that is consistently streaming (e.g. a Forex broker sends me the price changes of a currency). It's a GET request and it stays open.

First I tried to read the socket in the active mode with stream/2. But the error: it seems that stream/2 expects to have a full HTTP response for every incoming data. Mint said the HTTP status line was incorrect (:invalid_status_line) after the first successful read.

For this reason I switched to the passive mode with recv/3, to get data from the buffer without this check.
But I can't get the data in real-time from the buffer, there is a huge lag and I don't understand why.
For example I logged the following:

> buffer ~T[01:08:35.108000]: 
> [[] | "{\"type\":\"HEARTBEAT\",\"time\":\"2019-10-18T01:08:29.949199238Z\"}\n"]
> responses ~T[01:08:35.112000]:
> []

Note the "heartbeat" data is in the buffer at **01:08:35** above ^

> buffer ~T[01:08:40.083000]:
> []
> responses ~T[01:08:40.097000]:
> [
>   {:data, #Reference<0.1007427001.81264643.187480>,
>    "{\"type\":\"HEARTBEAT\",\"time\":\"2019-10-18T01:08:29.949199238Z\"}\n"}
> ]

I retrieve that data from the response **5 seconds late** at 01:08:40 ^
(and of course, in the context of the Forex, I need to get the data as soon as it is possible)

The code is trivial:

```
defp stream_response(conn) do
  {_, conn, responses} = HTTP.recv(conn, 0, :infinity)

  IO.inspect conn.request.data_buffer, label: "buffer " <> inspect Time.utc_now()

  IO.inspect responses, label: "responses " <> inspect Time.utc_now()

  stream_response(conn)
end
```

Also I do not understand why recv/3 returns me an empty list [] so often (as in the first log above).

A curl to that URL looks like that:

`curl -H "Authorization: Bearer <my token>" "https://stream-fxpractice.oanda.com/v3/accounts/<my account id>/pricing/stream?instruments=EUR_GBP"`

```
{"type":"HEARTBEAT","time":"2019-10-18T01:17:24.352078934Z"}
{"type":"HEARTBEAT","time":"2019-10-18T01:17:29.361340518Z"}
etcetera
```

All in all I just try to get the incoming data (one line per data and no empty lines) just as the output stream of the cURL above.

**Comment by ericmj** [comment]
> First I tried to read the socket in the active mode with stream/2. But the error: it seems that stream/2 expects to have a full HTTP response for every incoming data.

No, stream/2 continuously returns the response body as it comes in. If you get `:invalid_status_line` it suggests you forget to use the returned connection at some point, every time `conn` is returned you need to make sure to pass it to next mint function since it holds state about the connection that would be lost otherwise.

If you can show your code we might be able to help you more.

**Comment by thojanssens** [comment]
Thank you. I made sure to reuse the `conn` with `stream/2` but I got the same problem as with the passive mode, i.e. I retrieve data too late.

For example, the code below will log:

> buffer ~T[02:42:59.543000]: [[] | "{\"type\":\"HEARTBEAT\",\"time\":\"2019-10-19T02:42:58.523930708Z\"}\n"]
> responses ~T[02:42:59.545000]: []

Note the "heartbeat" data is in the buffer at the time 02:42:59. (`responses` will contain the empty list []. I don't understand why.)

> buffer ~T[02:43:04.559000]: []
> responses ~T[02:43:04.559000]: [
>   {:data, #Reference<0.2598560754.1930428418.27476>,
>    "{\"type\":\"HEARTBEAT\",\"time\":\"2019-10-19T02:42:58.523930708Z\"}\n"}
> ]

I effectively retrieve that heartbeat data from `responses` at 02:43:04, so 5 seconds later.

Code:

```
defmodule Forex do
  alias Mint.HTTP

  defp receive_next_and_stream(conn) do
    receive do
      message ->
        #IO.inspect(message, label: :message)
        result = Mint.HTTP.stream(conn, message)
        #IO.inspect result, label: :result

        {:ok, conn, responses} = result

        IO.inspect(conn.request.data_buffer, label: "buffer " <> inspect Time.utc_now())
        IO.inspect(responses, label: "responses " <> inspect Time.utc_now())

        receive_next_and_stream(conn)
    end
  end

  def stream_prices() do
    token = Application.fetch_env!(:forex, :token)
    account_id = Application.fetch_env!(:forex, :account_id)

    {:ok, conn} =
      HTTP.connect(:https, "stream-fxpractice.oanda.com", 443,
        #mode: :passive,
        timeout: :infinity
      )

    uri = "/v3/accounts/" <> account_id <> "/pricing/stream?instruments=EUR_GBP"

    {:ok, conn, request_ref} = HTTP.request(conn, "GET", uri, [{"Authorization", token}])

    receive_next_and_stream(conn)
  end
end
```

**Comment by whatyouhide** [comment]
@thojanssens do you have a snippet of code to reproduce this? Could you post everything that's printed out by that snippet of code, since the moment you connect?

**Comment by ericmj** [comment]
It can be an issue with how we handle chunked encoding but to determine that we need to look at the incoming data, please uncomment `#IO.inspect(message, label: :message)`.

**Comment by thojanssens** [comment]
I have sent the file with the credentials to your email. It's just the code above but with credentials.

Below are the logs since I connect (note the two last messages below: heartbeat data in the buffer at 03:36:20, and fetched in `responses` at 03:36:25)

```
iex> Forex.stream_prices

message: {:ssl,
 {:sslsocket, {:gen_tcp, #Port<0.7>, :tls_connection, :undefined},
  [#PID<0.221.0>, #PID<0.220.0>]},
 "HTTP/1.1 200 OK\r\nServer: openresty/1.9.15.1\r\nDate: Mon, 21 Oct 2019 03:36:12 GMT\r\nContent-Type: application/octet-stream\r\nTransfer-Encoding: chunked\r\nConnection: close\r\nAccess-Control-Allow-Header
s: Authorization, Content-Type, Accept-Datetime-Format, OANDA-Agent, ETag\r\nAccess-Control-Allow-Methods: PUT, PATCH, POST, GET, OPTIONS, DELETE\r\nAccess-Control-Allow-Origin: *\r\nAccess-Control-Expose-Heade
rs: ETag, RequestID\r\nRequestID: 763196637085578641\r\n\r\n10a\r\n"}

buffer ~T[03:36:15.705000]: [[] | ""]

responses ~T[03:36:15.717000]: [
  {:status, #Reference<0.2355769755.1311244290.16007>, 200},
  {:headers, #Reference<0.2355769755.1311244290.16007>,
   [
     {"server", "openresty/1.9.15.1"},
     {"date", "Mon, 21 Oct 2019 03:36:12 GMT"},
     {"content-type", "application/octet-stream"},
     {"transfer-encoding", "chunked"},
     {"connection", "close"},
     {"access-control-allow-headers",
      "Authorization, Content-Type, Accept-Datetime-Format, OANDA-Agent, ETag"},
     {"access-control-allow-methods", "PUT, PATCH, POST, GET, OPTIONS, DELETE"},
     {"access-control-allow-origin", "*"},
     {"access-control-expose-headers", "ETag, RequestID"},
     {"requestid", "763196627085878681"}
   ]}
]



message: {:ssl,
 {:sslsocket, {:gen_tcp, #Port<0.7>, :tls_connection, :undefined},
  [#PID<0.221.0>, #PID<0.220.0>]},
 "{\"type\":\"PRICE\",\"time\":\"2019-10-21T03:36:04.241876151Z\",\"bids\":[{\"price\":\"0.86401\",\"liquidity\":10000000}],\"asks\":[{\"price\":\"0.86423\",\"liquidity\":10000000}],\"closeoutBid\":\"0.86401\",
\"closeoutAsk\":\"0.86423\",\"status\":\"tradeable\",\"tradeable\":true,\"instrument\":\"EUR_GBP\"}\n\r\n"}

buffer ~T[03:36:15.730000]: [
  [[] | ""] |
  "{\"type\":\"PRICE\",\"time\":\"2019-10-21T03:36:04.241876151Z\",\"bids\":[{\"price\":\"0.86401\",\"liquidity\":10000000}],\"asks\":[{\"price\":\"0.86423\",\"liquidity\":10000000}],\"closeoutBid\":\"0.86401\"
,\"closeoutAsk\":\"0.86423\",\"status\":\"tradeable\",\"tradeable\":true,\"instrument\":\"EUR_GBP\"}\n"
]

responses ~T[03:36:15.735000]: []



message: {:ssl,
 {:sslsocket, {:gen_tcp, #Port<0.7>, :tls_connection, :undefined},
  [#PID<0.221.0>, #PID<0.220.0>]}, "3d\r\n"}

buffer ~T[03:36:20.659000]: []

responses ~T[03:36:20.661000]: [
  {:data, #Reference<0.2355769755.1311244290.16007>,
   "{\"type\":\"PRICE\",\"time\":\"2019-10-21T03:36:04.241876151Z\",\"bids\":[{\"price\":\"0.86401\",\"liquidity\":10000000}],\"asks\":[{\"price\":\"0.86423\",\"liquidity\":10000000}],\"closeoutBid\":\"0.86401\
",\"closeoutAsk\":\"0.86423\",\"status\":\"tradeable\",\"tradeable\":true,\"instrument\":\"EUR_GBP\"}\n"}
]



message: {:ssl,
 {:sslsocket, {:gen_tcp, #Port<0.7>, :tls_connection, :undefined},
  [#PID<0.221.0>, #PID<0.220.0>]},
 "{\"type\":\"HEARTBEAT\",\"time\":\"2019-10-21T03:36:17.443352878Z\"}\n\r\n"}

buffer ~T[03:36:20.676000]: [[] | "{\"type\":\"HEARTBEAT\",\"time\":\"2019-10-21T03:36:17.443352878Z\"}\n"]

responses ~T[03:36:20.677000]: []



message: {:ssl,
 {:sslsocket, {:gen_tcp, #Port<0.7>, :tls_connection, :undefined},
  [#PID<0.221.0>, #PID<0.220.0>]}, "3d\r\n"}

buffer ~T[03:36:25.720000]: []

responses ~T[03:36:25.721000]: [
  {:data, #Reference<0.2355769755.1311244290.16007>,
   "{\"type\":\"HEARTBEAT\",\"time\":\"2019-10-21T03:36:17.443352878Z\"}\n"}
]
```

**Comment by ericmj** [comment]
Thanks you for providing a way to reproduce and for the detailed output! 💜

**Comment by thojanssens** [comment]
No problem Eric. Will you make a release once the fix is done to get the dependency from mix?

I also had a quick question about the passive mode. It doesn't say in the doc why one would use the passive mode instead of the active mode. It mentions that passive mode is for special use cases without giving an example, so I don't really see; maybe I'm also still too much of a beginner to understand.

> By default Mint operates in active mode meaning that the process that started the connection receives socket messages. Mint also supports passive mode, where no messages are sent to the process and the process needs to fetch data out of the socket manually. The mode can be controlled at connection time through the :mode option in connect/4 or changed dynamically through set_mode/2. Passive mode is generally only recommended for special use cases.

**Comment by whatyouhide** [comment]
@thojanssens we will make a new release in a few days. As for the passive mode, we can improve the docs with examples but please open another issue about that if you can 🙃

**Comment by thojanssens** [comment]
Thank you for the release bro! ✌️

Curious: why did you deprecate `Mint.HTTP.request/4` that set the default body to `nil`?

**Comment by whatyouhide** [comment]
@thojanssens Mint is low level so we decided to deprecate a "courtesy" function and just require users to be explicit with `nil` as the body. This also opens up the possibility to have `request/6` with optional options `[]` as the last argument in case we want to support options for requests in the future. Hopefully that makes sense!

**Comment by thojanssens** [comment]
Makes total sense, that was a good decision.
